오픈마켓 II [STEP 2] Bonf, Finnn

finnn1 opened the pull request:
# 오픈마켓 II [STEP 2] PR 💸

---

안녕하세요! 쿠마(@leejun6694 ) Bonf, Finnn입니다 😊
오픈마켓 II [STEP 2] PR 보냅니다!
이번에도 잘 부탁드립니다!! 🙏

---

### 구현 내용

|**상품정보**|**상품수정**|**상품삭제**|
|:---:|:---:|:---:|
|![](https://i.imgur.com/hst1u1I.gif)|![](https://i.imgur.com/8j8CgFS.gif)|![](https://i.imgur.com/RKp7Pdr.gif)|


---

### 궁금한 점 🤷🏻‍♂️

> **이미지 파일 용량 조절 :**
> 프로젝트 요구사항에 업로드할 이미지의 파일 용량을 300kb로 제한하는 기능을 구현하도록 되어있었습니다.
> 저희의 한계로인해 이미지를 원하는 용량으로 바로 `compress` 할 수 있는 메서드를 찾지 못해, `.jpegData(compressionQuality:)` 메서드의 파라미터를 변경해가며 `while` 문을 통해 원하는 용량이 될 때까지 여러번 `compress` 하도록 구현했습니다.
> 
> 이러한 방법이 비효율적이라는 생각이 들어, 혹시 더 좋은 방법이 있는지 궁금합니다! ☺️🙏


---

### 고민한 점 🤔

#### CollectionView를 이용한 가로 스크롤, 페이징
> 처음 요구사항을 제대로 인지하지 못하여 ScrollView로 image를 나타냈습니다. CollectionView로 나타내려고 하는데, 가로 스크롤이 되지 않는 문제에 직면했습니다. ScrollView처럼 StackView를 내부에 넣어 활용하려하였지만 잘 되지 않았습니다.
 그 때 새로 알게된 부분이 `UICollectionViewCompositionalLayout`을 이용할 경우 `section.orthogonalScrollingBehavior` 값을 이용하여 조정할 수 있다는 것이었습니다. 저 값을 `.continues`으로 두었더니 가로 스크롤이 가능했지만, 페이징이 되지 않아 여러 사진이 겹치는 문제가 발생하여 `.groupPagingCentered`로 설정하여 해결했습니다.

---

#### 상품 삭제시 비밀번호 
> 서버로 `DELETE` 요청시, 사용자 `secret` 외에도 `product_secret`이 필요한 것을 확인했습니다. 삭제시 사용자에게 요청할 비밀번호를 `secret`으로 요청할지 `product_secret`으로 요청할지 고민한 끝에, `product_secret`은 사용자가 암기하기 어렵고, `secret`이 있으면 `product_secret`이 발급 가능하다는 점을 근거로 사용자 `secret`을 요청하는 방법을 채택했습니다.

> #### [POST] 상품 삭제 Secret 조회 Request
> |Name|Type|in|description|required|
> |:---:|:---:|:---:|:---:|:---:|
> |product_id|number|path|상품 식별 번호|✔︎|
> |identifier|string|header|판매자 식별 아이디|✔︎|
> |secret|string|body|판매자 비밀번호|✔︎|
> #### [DELETE] 상품 삭제 Request
> |Name|Type|in|description|required|
> |:---:|:---:|:---:|:---:|:---:|
> |identifier|string|header|판매자 식별 아이디|✔︎|
> |product_id|number|path|상품 식별 번호|✔︎|
> |product_secret|string|path|상품 삭제 secretKey|✔︎|

---

#### 데이터 Parsing시 프로퍼티 옵셔널처리
> 서버로 다양한 요청을 보내지만, `Response` 받는 `Data` 형식은 비슷하다는 것을 깨달았습니다.
> 다양한 `Response`를 하나의 `Decodable Type`으로 받을 수 있도록 프로퍼티들을 추가했지만, `images` 및 `description` 등 몇몇 `Response`에서는 사용되지 않는 프로퍼티에 옵셔널 처리를 해줌으로써 코드의 유연성 및 재사용성을 높혔습니다.

finnn1 committed:
refactor: 불필요한 구문 제거 및 네이밍 수정

finnn1 committed:
feat: ProductsDetailView 레이아웃 구성

apwierk2451 committed:
feat: navigation bar button 구현

apwierk2451 committed:
feat: delegate를 이용하여 정보 전달

- ProductsViewController -> ProductsDetailViewController

finnn1 committed:
refactor: 상품등록 페이지로 이동시 Title 추가

finnn1 committed:
fix: Post시 discounted_price가 들어가지 않는 현상 수정

finnn1 committed:
fix: 금액이 제대로 뜨지 않는 현상 수정

apwierk2451 committed:
feat: detailView에서 이미지가 모두 출력되도록 구현

finnn1 committed:
feat: 상품수정 페이지에서 patch 기능 구현

finnn1 committed:
fix: patch 후 completion 실행안되는 문제 수정

finnn1 committed:
fest: getProductSecret 메서드 구현

apwierk2451 committed:
feat: 상품삭제 선택할 경우 DELETE 기능 구현

apwierk2451 committed:
refactor: detailView에 가격 Label 수정

- currency 표시
- strikeThrough 구현

finnn1 committed:
feat: 상품 이미지를 띄우기 위한 CollectionView 구성

finnn1 committed:
refactor: CollectionVIew 페이징 구현

finnn1 committed:
feat: Image 파일 사이즈 변환 기능 구현

apwierk2451 committed:
refactor: view를 교체하는 방식에서 view를 덮는 방식으로 변경

apwierk2451 committed:
refactor: 코드 컨벤션 수정

- struct 파일로 나눔
- 기본 값 제거
- 중복 기능 융합

finnn1 committed:
refactor: 상품수정 페이지 진입시 Currency 할당되도록 구현

finnn1 committed:
refactor: 선택했던 셀 선택해제 되도록 구현

apwierk2451 committed:
refactor: 강제 언래핑 제거

apwierk2451 committed:
refactor: 은닉화

finnn1 committed:
style: 전체적인 컨벤션 수정

apwierk2451 committed:
feat: 이미지 터치 시 수정, 삭제 액션 시트 추가

finnn1 committed:
refactor: DELETE 관련 메서드 분리

finnn1 committed:
feat: IdentifierError 타입 정의

finnn1 committed:
refactor: HTTP Request 요청 실패시 사용자에게 알림

leejun6694 commented on OpenMarket/OpenMarket/Controllers/ProductsDetailViewController.swift:
```diff
@@ -2,246 +2,144 @@ import UIKit
-        guard let userInfo = notification.userInfo,
-              let keyboardFrame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else {
-            return
+        guard let delegate = delegate else { return }
```
method 를 따로 만들어주면 어떨까요?

finnn1 replied:
넵 별도의 메서드로 분리해주었습니다! ☺️

leejun6694 commented on OpenMarket/OpenMarket/Controllers/ProductsRegistViewController.swift:
```diff
@@ -0,0 +1,301 @@
+        ])
+    }
+    
+    private func presentAlertMessage(message: String) {
```
이 method 를 포함해서, 여러 코드들이 `ProductsDetailViewController` 와 중복되고 있습니다.
중복 코드를 줄일 수 있는 방법에 대해 생각해볼까요?

finnn1 replied:
해당 메서드는 프로토콜로 따로 분리해주었습니다 ☺️

leejun6694 commented on OpenMarket/OpenMarket/Extension/Data+Extension.swift:
```diff
@@ -0,0 +1,15 @@
+extension Data {
+    mutating func append(_ string: String) {
+        if let data = string.data(using: .utf8) {
+            self.append(data)
```
`append` 못한 경우는 구현부에 따로 알리지 않아도 되는걸까요?
에러 처리라던가..

finnn1 replied:
해당 부분은 에러처리로 빼주었습니다 🙏

leejun6694 commented on OpenMarket/OpenMarket/Models/DecodableTypes/Image.swift:
```diff
@@ -0,0 +1,15 @@
+struct Image: Decodable, Hashable {
```
너무 일반적인 이름이네요..!
`ProductImage` 와 같은 이름은 어떨까요?

finnn1 replied:
타입명을 `ProductImage`로 변경해주었습니다 ☺️

leejun6694 commented on OpenMarket/OpenMarket/Models/Enum/Currency.swift:
```diff
@@ -0,0 +1,19 @@
+enum Currency: Int, Decodable, CaseIterable {
```
`rawValue` 를 `String` 으로 하면 더 복잡해질까요?
toIndex 에서 default 값이 생기기도 하고, uppercased 가 계속 쓰이고 있어서, KRW, USD 를 rawValue 로 가져가면 어떨까 생각해봤습니다 : )

finnn1 replied:
`rawValue`를 `String`으로 변경해주었습니다 🙏👍

leejun6694 commented on OpenMarket/OpenMarket/Models/Enum/DetailViewTitle.swift:
```diff
@@ -0,0 +1,4 @@
+enum DetailViewTitle: String {
+    case regist = "상품등록"
```
case 의 특징을 사용해야하는게 아니라면, static let 은 어떨까요?

finnn1 replied:
`case` 대신, `static let` 으로 변경해주었습니다! ☺️

leejun6694 commented on OpenMarket/OpenMarket/Models/Error/IdentifierError.swift:
```diff
@@ -0,0 +1,3 @@
+enum IdentifierError: Error {
+    case NotYourProduct
```
case 는 소문자로 시작합니다.

finnn1 replied:
앗 그러네요..! 미처 발견하지 못한 것 같습니다 😢
`lowerCamelCase` 형태로 변경해주었습니다 🙏

leejun6694 commented on OpenMarket/OpenMarket/Models/ProductsDataManager.swift:
```diff
@@ -1,47 +1,5 @@
 import UIKit.NSDataAsset
```
여전히 필요한가요..!

leejun6694 commented on OpenMarket/OpenMarket/Models/ProductsDataManager.swift:
```diff
@@ -117,8 +96,41 @@ struct ProductsDataManager: Decodable {
+        
+        let task = URLSession.shared.dataTask(with: postRequest) { (data, response, error) in
+            if let response = response {
+                print(response)
```
print..!

finnn1 replied:
`completion` 으로 `result` 타입을 전달해주는 것으로 변경했습니다! ☺️
감사합니다 🙏

leejun6694 reviewed (changes requested):
Bonf, Finnn 안녕하세요~
이번 pr 은 수정사항이 많았네요!
-
다른 구현도 공부할 부분이 많지만, 이번 pr 에서는 `중복된 코드나 기능들을 어떻게 묶고 공통으로 만들 수 있을지` 고민해봤으면 합니다.
- 중복되는 코드나 기능 조사
- 공통 코드로 만들 필요성 고민
- 어떻게 공통 코드로 묶을지 고민
- 구현

혹시나 중간에 궁금한 점이 있다면, 언제든 말씀 주세요~~!

leejun6694 commented:
> 이미지 파일 용량 조절

음.. 저도 while 과 같은 반복문을 통해 구현할 것 같아요. 음.... 이진탐색으로도 가능할지는 시도해볼거 같네요ㅎㅎ

finnn1 committed:
doc: OpenMarket II [Step 2] README.md 추가

finnn1 committed:
doc: Readme 참고문헌 내용 추가

finnn1 committed:
Update README.md

finnn1 committed:
Update README.md

finnn1 committed:
Update README.md

finnn1 committed:
Update README.md

apwierk2451 committed:
Update README.MD

- 목차 id 수정

apwierk2451 committed:
Update README.md

apwierk2451 committed:
refactor: 네이밍 변경

apwierk2451 committed:
refactor: Data append 에러처리 기능 구현

apwierk2451 committed:
refactor: Alert메세지 프로토콜로 기능 구현

apwierk2451 committed:
refactor: enum 정보 수정

- Currency: Int -> String, toIndex()추가
- DetailViewTitle: case -> static let 으로 구현

leejun6694 commented on OpenMarket/OpenMarket/Models/AlertMessage.swift:
```diff
@@ -0,0 +1,15 @@
+    func presentAlertMessage(controller: UIViewController, message: String)
+}
+
+extension AlertMessage {
```
`where` 를 활용해볼까요?

leejun6694 replied:
`where` 를 활용해볼까요?

leejun6694 commented on OpenMarket/OpenMarket/Models/AlertMessage.swift:
```diff
@@ -0,0 +1,15 @@
+import UIKit.UIAlertController
+
+protocol AlertMessage {
```
👍 👍 👍

leejun6694 commented on OpenMarket/OpenMarket/Controllers/ProductsDetailViewController.swift:
```diff
@@ -1,247 +1,139 @@
     
-    func isEmptySaleAndStockTextField(_ textField: UITextField) -> Bool {
-        guard let detailView = view as? ProductDetailView else { return false }
+    private func deleteProcess() {
```
`AlertMessage` 를 잘 활용하면, 다른 alert 들도 필요한 값들만 넘겨줘서 사용할 수 있지 않을까요!?

leejun6694 commented on OpenMarket/OpenMarket/Controllers/ProductsRegistViewController.swift:
```diff
@@ -0,0 +1,291 @@
+        let changeImageAction = UIAlertAction(title: "교체", style: .default) { [weak self] action in
+            guard let self = self else { return }
+            
+            print("이미지 교체 버튼 눌림")
```
print...! ㅎㅎ

leejun6694 commented on OpenMarket/OpenMarket/Models/AlertMessage.swift:
```diff
@@ -0,0 +1,15 @@
+        
+        alert.addAction(action)
+        controller.present(alert, animated: true)
+    }
```
기왕 만들어본김에 좀 더 공용 코드로 어떻게 사용할 수 있을까요?
버튼이 하나만 있는 경우? 두개인 경우? textField 가 있는 경우? action 이 필요한 경우?

leejun6694 commented on OpenMarket/OpenMarket/Models/Enum/Currency.swift:
```diff
@@ -0,0 +1,26 @@
+    }
+    
+    
+    
```
메모리가 낭비되고 있어요!

leejun6694 reviewed (approved):
리뷰 내용을 잘 반영해주셨네요~~!
좀 더 생각해볼 부분이 있어서, 코멘트 남겼습니다 : )
